src/library_manager/lib_manager.c

```diff
@@ -68,114 +68,133 @@ static int lib_manager_load_data_from_storage(void __sparse_cache *vma, void *s_
+		if (!mod->segment[idx].flags.r.load)
+			continue;
 
-	/* Copy Code */
-	ret = lib_manager_load_data_from_storage(va_base_text, src_txt, st_text_size,
-						 SYS_MM_MEM_PERM_RW | SYS_MM_MEM_PERM_EXEC);
-	if (ret < 0)
-		goto err;
+		va_base = UINT_TO_POINTER(mod->segment[idx].v_base_addr);
+		size = mod->segment[idx].flags.r.length * PAGE_SZ;
+		ret = sys_mm_drv_unmap_region(va_base, size);
+		if (ret < 0)
+			return ret;
+	}
 
-	/* Copy RODATA */
-	ret = lib_manager_load_data_from_storage(va_base_rodata, src_rodata,
-						 st_rodata_size, SYS_MM_MEM_PERM_RW);
-	if (ret < 0)
-		goto err;
+	return 0;
+}
 
-	/* There are modules marked as lib_code. This is code shared between several modules inside
-	 * the library. Load all lib_code modules with first none lib_code module load.
-	 */
-	if (!mod->type.lib_code)
-		ext_lib->mods_exec_load_cnt++;
-
-	if (ext_lib->mods_exec_load_cnt == 1) {
-		struct sof_man_module *module_entry =
-		    (struct sof_man_module *)((char *)desc + SOF_MAN_MODULE_OFFSET(0));
-		for (size_t idx = 0; idx < desc->header.num_module_entries;
-				++idx, ++module_entry) {
-			if (module_entry->type.lib_code) {
-				ret = lib_manager_load_module(lib_id << LIB_MANAGER_LIB_ID_SHIFT |
-							      idx, mod, desc);
-				if (ret < 0)
-					goto err;
-			}
+/* There are modules marked as lib_code. This is code shared between several modules inside
+ * the library. Load all lib_code modules with first none lib_code module load.
+ */
+static int lib_manager_load_libcode_modules(const uint32_t module_id,
+					 const struct sof_man_fw_desc *const desc)
+{
+	struct ext_library *const ext_lib = ext_lib_get();
+	const struct sof_man_module *module_entry = (struct sof_man_module *)
+		((char *)desc + SOF_MAN_MODULE_OFFSET(0));
+	const uint32_t lib_id = LIB_MANAGER_GET_LIB_ID(module_id);
+	int ret, idx;
+
+	if (++ext_lib->mods_exec_load_cnt > 1)
+		return 0;
+
+	for (idx = 0; idx < desc->header.num_module_entries; ++idx, ++module_entry) {
+		if (module_entry->type.lib_code) {
+			ret = lib_manager_load_module(lib_id << LIB_MANAGER_LIB_ID_SHIFT | idx,
+						      module_entry);
+			if (ret < 0)
+				goto err;
 		}
 	}
 
 	return 0;
 
 err:
-	sys_mm_drv_unmap_region((__sparse_force void *)va_base_text, st_text_size);
-	sys_mm_drv_unmap_region((__sparse_force void *)va_base_rodata, st_rodata_size);
+	for (--idx, --module_entry; idx >= 0; --idx, --module_entry) {
+		if (module_entry->type.lib_code) {
+			ret = lib_manager_unload_module(module_entry);
+			if (ret < 0)
+				goto err;
+		}
+	}
 
 	return ret;
 }
 
-static int lib_manager_unload_module(uint32_t module_id, struct sof_man_module *mod,
-				     struct sof_man_fw_desc *desc)
+/* There are modules marked as lib_code. This is code shared between several modules inside
+ * the library. Unload all lib_code modules with last none lib_code module unload.
+ */
+static int lib_manager_unload_libcode_modules(const uint32_t module_id,
```

Comment:
let's leave this out until we have a test case

Reply:
I think CI will test this ? Btw , I dont see this as a blocker if missing from CI today.

Reply:
@lgirdwood CI will only test this if a test is created? This is a particular functionality that I'm not sure if we get a user in the next year or two or... And without a single user this will just make reviewing and maintaining more difficult and increase the code size needlessly.

Reply:
There will be Windows tests for this feature and others not targeting Linux..

Reply:
@softwarecki whats the ETA for client code that uses the extended/unused APIs today? I think we can quickly solve this with a KCONFIG that builds the unload logic when selected (and we can fully enable when client code is upstream). e.g. KCONFIG_LIB_MANAGER_UNLOAD.

Reply:
@lgirdwood a `CONFIG_...` option would help but not for unloading - this is just one function from the API in question: it's about both `lib_manager_load_libcode_modules()` and `lib_manager_unload_libcode_modules()`.